sno/working_copy/postgis.py

```diff
@@ -336,14 +249,52 @@ def delete(self):
             dbcur.execute(SQL("DROP SCHEMA {} CASCADE").format(Identifier(self.schema)))
 
     def write_meta(self, dataset):
-        pass
+        """
+        Populate the following postgis tables with data from this dataset:
+        geometry_columns, spatial_ref_sys
+        """
+        table = dataset.table_name
+        with self.session() as db:
+            dbcur = db.cursor()
+            dbcur.execute(
+                SQL(
+                    "DELETE FROM {} WHERE f_table_schema=%s AND f_table_name=%s;"
+                ).format(self._table_identifier("geometry_columns")),
+                (self.schema, table),
+            )
 
-    def read_meta(self, dataset):
-        return {}
+            geom_columns = postgis_adapter.generate_postgis_geometry_columns(
+                dataset, self.schema
+            )
+            for row in geom_columns:
+                dbcur.execute(
+                    SQL("INSERT INTO {} ({}) VALUES ({});").format(
+                        self._table_identifier("geometry_columns"),
+                        SQL(",").join([Identifier(k) for k in row]),
+                        SQL(",").join([SQL("%s")] * len(row)),
+                    ),
+                    tuple(row.values()),
+                )
 
-    def save_config(self, **kwargs):
-        self.repo.config["sno.workingcopy.version"] = 1
-        self.repo.config["sno.workingcopy.path"] = self.uri
+            spatial_refs = postgis_adapter.generate_postgis_spatial_ref_sys(dataset)
+            for row in spatial_refs:
+                dbcur.execute(
+                    SQL(
+                        """
+                        INSERT INTO {} ({}) VALUES ({})
+                        ON CONFLICT (srid) DO UPDATE
```

Comment:
I think we might need to FIXME this... If some repo has a messed up definition of a CRS code, should it really be able to override the one that PostGIS has installed, potentially messing up all transforms happening across the DB?

Maybe for now we warn about what we're overwriting?

Reply:
For now, I've just stopped writing CRS's to the `spatial_ref_sys` table. The tests still pass: everything works as expected as long as no dataset contains any custom CRS that the `spatial_ref_sys` doesn't know about, and I don't yet have any postgres tests that try to set a custom CRS. I'll do it in a future PR, I've added a TODO.

We have some options - for instance, we could decide never to overwrite anything in the range 0-9999. If our CRS would overwrite one of those, we could add 1000000 to it (or whatever) - so creating our own range of "CRS that have an authority ID in the range 0-9999 but don't match Postgis definitions", with SRIDs in the range 1000000 to 1009999

Of course, we could still have collisions within that namespace too.... unless we use a larger range based on a hash of the CRS.

And we could do some work on deciding whether or not CRS definitions match and so if they both need a separate ID, or if they can share an ID in spite of minor differences. Eg using the SpatialRef functions for finding matching definitions, or by comparing proj4 text.

So many options...

Reply:
So the EPSG database is (by design) not mutable — if a coordinate transform  parameter is wrong they issue a new number and deprecate the old one. (names/areas-of-use/etc might change, but not the actual math parameters)

I'd suggest as a reasonably pragmatic first step:

1. INSERT missing CRS entries into `spatial_ref_sys`
2. Don't update existing CRS entries if they exist already

Call it done there, and we can revisit further down the line.

I guess this is the sort of thing a `fsck` command could check in detail?

Reply:
with the above approach, the solution for something weird happening for an end user becomes either:

1. delete the row from `spatial_ref_sys` and Sno will fix it
2. add a row to `spatial_ref_sys` with the correct settings and Sno will respect it